category_theory/products.lean

```diff
@@ -0,0 +1,106 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Stephen Morgan, Scott Morrison
+import .functor_category
+
+namespace category_theory
+
+universes u₁ v₁ u₂ v₂ u₃ v₃ u₄ v₄
+
+section
+variables (C : Type u₁) [category.{u₁ v₁} C] (D : Type u₂) [category.{u₂ v₂} D]
+
+/--
+`product_category C D` gives the cartesian product of two categories.
+-/
+instance product_category : category.{(max u₁ u₂) (max v₁ v₂)} (C × D) := 
+{ Hom     := λ X Y, ((X.1) ⟶ (Y.1)) × ((X.2) ⟶ (Y.2)),
+  id      := λ X, ⟨ 𝟙 (X.1), 𝟙 (X.2) ⟩,
+  comp    := λ _ _ _ f g, (f.1 ≫ g.1, f.2 ≫ g.2),
+  id_comp := begin  /- `obviously'` says: -/ intros, cases X, cases Y, cases f, dsimp at *, simp end,
+  comp_id := begin /- `obviously'` says: -/ intros, cases X, cases Y, cases f, dsimp at *, simp end,
+  assoc   := begin /- `obviously'` says: -/ intros, cases W, cases X, cases Y, cases Z, cases f, cases g, cases h, dsimp at *, simp end }     
+end 
+
+namespace product_category
+
+section -- rfl lemmas for product_category
+variables {C : Type u₁} [𝒞 : category.{u₁ v₁} C] {D : Type u₂} [𝒟 : category.{u₂ v₂} D]
+include 𝒞 𝒟
+
+@[simp,ematch] lemma id (X : C) (Y : D) : 𝟙 (X, Y) = (𝟙 X, 𝟙 Y) := rfl
+@[simp,ematch] lemma comp {P Q R : C} {S T U : D} (f : (P, S) ⟶ (Q, T)) (g : (Q, T) ⟶ (R, U)) : f ≫ g = (f.1 ≫ g.1, f.2 ≫ g.2) := rfl
+end
+
+section 
+variables (C : Type u₁) [𝒞 : category.{u₁ v₁} C] (D : Type u₁) [𝒟 : category.{u₁ v₁} D]
+include 𝒞 𝒟
+
+/-- 
+`product_category.uniform C D` is an additional instance specialised so both factors have the same universe levels. This helps typeclass resolution.
+-/
+instance uniform : category.{u₁ v₁} (C × D) := category_theory.product_category C D
+end
+
+-- Next we define the natural functors into and out of product categories. For now this doesn't address the universal properties.
+
+/-- `right_injection_at C Z` is the functor `X ↦ (X, Z)`. -/
+definition right_injection_at (C : Type u₁) [category.{u₁ v₁} C] {D : Type u₁} [category.{u₁ v₁} D] (Z : D) : C ↝ (C × D) := 
```

Comment:
The last two should be called `fst` and `snd`. For the first two I don't think we've given these names, but I guess we could borrow `inl` and `inr` from the coproduct terminology. (*Note:* `right_injection_at` would be `inl` from that terminology!) I think that when possible we should borrow names used for analogous notions in the lambda calculus.